### ISSUE TITLE
Deserialize issue when a singleton default is set on a backing class

### ISSUE THREAD

**Issue by jroyals** [JamesNK/Newtonsoft.Json#2500]
I team, I've found a strange issue with JSON parsing, discovered using the CosmosDB .NET SDK v3.16.0 but I've isolated it to Newtonsoft.Json v12.0.3 on .NET Core 5.

May be related to https://github.com/JamesNK/Newtonsoft.Json/issues/2451

Easiest to explain with a bit of code.

### Example xUnit code

```csharp
using Newtonsoft.Json;
using Xunit;
using Xunit.Abstractions;

namespace Storefront.API.IntegrationTests
{
    public class JsonParseIssueTest
    {
        private readonly ITestOutputHelper _testOutputHelper;
        
        public JsonParseIssueTest(ITestOutputHelper testOutputHelper)
        {
            _testOutputHelper = testOutputHelper;
        }

        [Fact]
        public void ShowJsonParseIssue()
        {
            var json = @"
                {
                  'foo': {
                    'widgetCount': 91,
                    'widgetAvailable': false
                  },
                  'bar': {
                    'widgetCount': 1455,
                    'widgetAvailable': true
                  }
                }";
            var o = JsonConvert.DeserializeObject<TestOuter>(json);

            _testOutputHelper.WriteLine(o.Foo.ToString()); // writes 'TestSubNested { WidgetCount = 1455, WidgetAvailable = True }' <-- OOPS!
            _testOutputHelper.WriteLine(o.Bar.ToString()); // writes 'TestSubNested { WidgetCount = 1455, WidgetAvailable = True }'
        }
    }

    public record TestOuter
    {
        public TestInner Foo { get; init; } = TestInner.NullValue; // NOTE singleton defaults
        public TestInner Bar { get; init; } = TestInner.NullValue;  
    }

    public record TestInner
    {
        public static readonly TestInner NullValue = new TestInner {WidgetAvailable = false, WidgetCount = 0};
        public decimal WidgetCount { get; init; }
        public bool WidgetAvailable { get; init; }
    }
}
```

### Expected behavior

After deserialization the the two console logs should reflect the actual values of the records they populate.

### Actual behavior

The console logs print out the same value - whatever the last item is (see comments in code example above).

### Workaround

If I remove the default value set on `TestOuter` then it works as expected. Similarly, if I inline the singleton and create two separate (but equal) defaults on TestOuter then the problem also resolves.

**Comment by elgonzo** [comment]
Instruct the deserializer to _not_ reuse existing object instances (default mode of operation of the deserializer is to reuse existing instances). https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_JsonSerializerSettings_ObjectCreationHandling.htm

    var o = JsonConvert.DeserializeObject<TestOuter>(
        json,
        new JsonSerializerSettings { ObjectCreationHandling = ObjectCreationHandling.Replace }
    );

**Comment by jroyals** [comment]
Yep I see what's happened, I added another print statement showing the value of `TestInner.NullValue` and it appears to have mutated the property values, so the refences were never updated. The config change to replace an object does indeed resolve it.

Since `NullValue` is meant to be immutable (no setters, only an getter and init) then something going in and mutating the values underneath wasn't what I expected. Perhaps "Auto" handling should consider the mutability of the object it is working with to make a replace/reuse decision?

**Comment by elgonzo** [comment]
Considering init-only setters were only introduced with C# 9.0, it would be surprising if Newtonsoft.Json v12.0.3 would be able to identify these, considering Newtonsoft.Json v12.0.3 was released much earlier around the end of 2019.

I don't know whether the 13.0.x betas or the master/development branch of  Newtonsoft.Json already support identifiying init-only properties or records, though.


If not, you might try employing a custom ContractResolver that inspects properties of a type for having a setter with a custom modifier `System.Runtime.CompilerServices.IsExternalInit` (see this SO answer for some details: https://stackoverflow.com/questions/64749277/in-c9-how-do-init-only-properties-differ-from-read-only-properties), and sets `JsonProperty.Readable` (or was that `JsonProperty.Writable`?) accordingly. I don't know if there are already some custom ready-to-use ContractResolver implementations out there that do exactly that thing, but if not, writing your own ContractResolver shouldn't be too difficult. Basically, the init-only setter check on a property boils down to (warning, ugly code):

    var propSetter = propertyInfo.GetSetMethod();
    var isPropertyReadOnly =
        propSetter == null
        || propSetter.ReturnParameter.GetRequiredCustomModifiers().Contains(typeof(System.Runtime.CompilerServices.IsExternalInit));
